msal/src/Microsoft.Identity.Client/TokenCache.cs

```diff
@@ -684,15 +688,22 @@ internal MsalAccountCacheItem GetAccountCacheItem(MsalAccountCacheKey msalAccoun
         {
             InstanceDiscoveryMetadataEntry instanceDiscoveryMetadata = null;
 
+            Uri authorityHost = new Uri(authority);
             var authorityType = Authority.GetAuthorityType(authority);
-            if (authorityType == Core.Instance.AuthorityType.Aad || authorityType == Core.Instance.AuthorityType.B2C)
+            if (authorityType == Core.Instance.AuthorityType.Aad || 
+                authorityHost.Host.Equals(MicrosoftLogin, StringComparison.OrdinalIgnoreCase))
             {
                 instanceDiscoveryMetadata = await AadInstanceDiscovery.Instance.GetMetadataEntryAsync(
                     HttpManager,
                     TelemetryManager,
                     new Uri(authority),
                     validateAuthority,
                     requestContext).ConfigureAwait(false);
+                return instanceDiscoveryMetadata;
+            }
+            if (authorityType == Core.Instance.AuthorityType.B2C)
+            {
+                return instanceDiscoveryMetadata;
             }
             return instanceDiscoveryMetadata;
```

Comment:
why not just return null for these 2 cases. only for AAD will it be set to anything besides null

Reply:
when i tested w/login.microsoftonline.com, the cache key was login.windows.net (or something like that), so we need to do the instance discovery for the login.microsoftonline.com authorities, but not the b2clogin.com ones

---
In reply to: [234387980](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#discussion_r234387980) [](ancestors = 234387980)

Reply:
not what I meant. You still  have the instance discovery happening for the AAD case above. You are actuvely not doing the instance discovery here. This place the variable will always be null.

---
In reply to: [234388128](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#discussion_r234388128) [](ancestors = 234388128,234387980)

Reply:
oh i get it.

---
In reply to: [234388879](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#discussion_r234388879) [](ancestors = 234388879,234388128,234387980)